index.bs

```diff
@@ -7260,6 +7260,76 @@ The [=remote end steps=] given |session|, and |command parameters| are:
 
 </div>
 
+## Emulation ## {#module-emulation}
+
+The <dfn export for=modules>emulation</dfn> module contains functionality for
+emulating device-dependent and system-dependent parameters.
+
+### Definition ### {#module-emulation-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+EmulationCommand = (
+  emulation.SetViewport
+)
+</pre>
+
+### Types ### {#module-emulation-types}
+
+### Commands ### {#module-emulation-commands}
+
+#### The emulation.setViewport Command ####  {#command-emulation-setViewport}
+
+The <dfn export for=commands>emulation.setViewport</dfn> command turns the emulation of the specified viewport on.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+      emulation.SetViewport = {
+        method: "emulation.setViewport",
+        params: emulation.SetViewportParameters
+      };
+
+      emulation.SetViewportParameters = {
+        context: browsingContext.BrowsingContext,
+        ? viewport: emulation.Viewport
+      }
+
+      emulation.Viewport = {
+        ? width: js-uint,
+        ? height: js-uint,
+      }
+    </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl">
+     EmptyResult
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for emulation.setViewport">
+
+The [=remote end steps=] with |command parameters| are:
+
+1. Let |context id| be the value of the <code>context</code> field of
+   |command parameters|.
+
+1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+   with |context id|.
+
+1. Let |viewport| be value of the <code>viewport</code> field of |command parameters| in [[webdriver#dfn-css-pixels]].
+
+1. Apply |viewport| as the actual vieport as defined in [[css-viewport#the-viewport]] to the |context| and all current and future sub-contexts that have a concept of a viewport.
```

Comment:
So with the understanding that the viewport spec is still in flux; I don't think this is unambiguous enough to be confident of interoperable implementations. Which viewport (initial / actual / layout) is being adjusted? What does it mean to set it to a Infra Map (you probably want to set individual fields instead). What's a "sub-context"? It seems inevitable that this is inherited into iframes, but is the idea that it's also inherited into auxillary browsing contexts (e.g. from `window.open`)? What about random new browsing contexts (e.g. from the `browsingContext.create` command)? Should there be a way to set the viewport properties globally so it applies to all contexts?

Reply:
The PR is not ready for review yet but I can provide some answers now, thanks for the good questions and suggestions!

> Which viewport (initial / actual / layout) is being adjusted? 
Currently, I believe that actual viewport is the one that needs to adjusted.

> What does it mean to set it to a Infra Map (you probably want to set individual fields instead)

I didn't get the question about the infra map? what is an infra map?

> It seems inevitable that this is inherited into iframes, but is the idea that it's also inherited into auxillary browsing contexts (e.g. from window.open)? 

I think it should not apply to iframes because they don't really have a viewport concept (dimensions of the iframe are controlled by the layout). Auxillary browsing contexts should be excluded too. By sub-contexts that have a concept of a viewport I mean something like fenced frames, prerendered or bfcache contexts that the current context might navigate to but which exist in parallel to the current context. I am still searching for a good term and if such details really need to be defined.

> What about random new browsing contexts (e.g. from the browsingContext.create command)? Should there be a way to set the viewport properties globally so it applies to all contexts?

For now, I think there is no use case for that. Perhaps it would make sense to specify the default viewport via capabilities? For example, it can be done via CLI flag in Chrome.

Reply:
> Perhaps it would make sense to specify the default viewport via capabilities? For example, it can be done via CLI flag in Chrome.

Can you clarify which cli flag you have in mind here? AFAIK there is no way to set viewport from cli (unless you mean window size, but that includes the browser's UI so IMHO it's not a good candidate for a cross browser API).

Reply:
Infra maps: https://infra.spec.whatwg.org/#maps It's a spec-internal key/value pair type which is how we represent command bodies in the spec prose. So in your case, you first need to check if `|command parameters| [=map/contains=] "<code>viewport</code>` and if it does `|command parameters["<code>viewport</code>"]` will be another map that (maybe) has properties named `"<code>width</code>"` and `"<code>height</code>"`. That's not an object you can work with directly; you'd need to extract the individual fields.

> By sub-contexts that have a concept of a viewport I mean something like fenced frames, prerendered or bfcache contexts that the current context might navigate to but which exist in parallel to the current context.

So this seems like a testable question: if I navigate to A, then navigate to B such that A ends up in bfcache, then set the viewport size, then go back in history to A, do I end up with the new viewport size or the previous one (I assume that for a normal navigation the viewport size won't change?). I think that's something that needs to be clear.

Reply:
@ewinslow you are right, my memory failed me :) the flag does not exist and, in Puppeteer's case, it sets the default viewport after the page is created (`defaultViewport` is a launch argument).

Reply:
I have updated the spec text which I believe should clarify the questions. 

> Should there be a way to set the viewport properties globally so it applies to all contexts?

Currently, we propose only changing the emulation per context.